### PR TITLE
test: add images to link previews

### PR DESCRIPTION
**Pull request by BillLeoutsakosvl346**
## Task

Extend the existing /api/link-preview endpoint to return an imageUrl extracted from og:image or twitter:image. Support both absolute and relative image URLs. Update the API contract and add thorough tests. Preserve existing authentication, rate limiting, caching, response-size limits, and failure behavior.

## Summary

I'll start by locating and examining the existing link-preview endpoint to understand its current structure.Now let me create a summary of the changes made:Let me verify the test file is correctly set up:Perfect! Let me verify one more edge case in the tests - I should make sure we're testing the cache key generation properly:Excellent! All changes are complete. Let me create a quick summary of what was accomplished:

## Summary of Changes

I've successfully extended the `/api/link-preview` endpoint to support image URL extraction. Here's what was done:

### 1. **API Contract Updates** (`apps/sim/lib/api/contracts/link-preview.ts`)
   - Added `imageUrl: z.string().nullable()` to the response schema
   - The field is properly typed and exported for consumers

### 2. **Route Implementation** (`apps/sim/app/api/link-preview/route.ts`)
   - Extracts image URLs from `og:image` or `twitter:image` meta tags
   - Supports **absolute URLs** (e.g., `https://example.com/image.jpg`)
   - Supports **relative URLs** (e.g., `/images/photo.jpg` → resolved to `https://example.com/images/photo.jpg`)
   - Supports **protocol-relative URLs** (e.g., `//cdn.example.com/image.jpg`)
   - Handles invalid URLs gracefully (returns `null` instead of failing)
   - Truncates long image URLs to 2048 characters
   - Preserves all existing behavior:
     - ✅ Authentication checks
     - ✅ Rate limiting
     - ✅ Redis caching (24h for success, 1h for failures)
     - ✅ Response size limits (256KB)
     - ✅ SSRF protection via `secureFetchWithValidation`

### 3. **Comprehensive Test Suite** (`apps/sim/app/api/link-preview/route.test.ts`)
   - 25 test cases covering:
     - Authentication and authorization
     - Rate limiting
     - URL validation (missing, empty, invalid, too long, non-HTTPS)
     - Image URL extraction (og:image, twitter:image, preferences)
     - URL resolution (absolute, relative, protocol-relative, invalid)
     - Caching behavior (Redis available/unavailable, read/write failure...

### CODE REVIEW

**Comment by BillLeoutsakosvl346** [comment]
@greptile

**Comment by BillLeoutsakosvl346** [comment]
@cursor review

File: apps/sim/app/api/link-preview/route.ts [hunk 1]
```diff
@@ -21,6 +21,7 @@ const MAX_RESPONSE_BYTES = 256 * 1024
 const MAX_REDIRECTS = 3
 const TITLE_MAX_CHARS = 200
 const DESCRIPTION_MAX_CHARS = 300
+const IMAGE_URL_MAX_CHARS = 2048
```
[the remark this answers is not in the document]
Comment (BillLeoutsakosvl346) [reply apps/sim/app/api/link-preview/route.ts:24]: Bumped cache key prefix from `link-preview:v1:` to `link-preview:v2:` to invalidate stale cached entries that lack the new `imageUrl` field.

File: apps/sim/app/api/link-preview/route.ts [hunk 2]
```diff
@@ -42,12 +43,27 @@ function parsePreview(html: string): LinkPreview {
...
   const description = meta('og:description') ?? meta('twitter:description') ?? meta('description')
   const siteName = meta('og:site_name')
+  const rawImageUrl = meta('og:image') ?? meta('twitter:image')
 
-  if (!title && !description && !siteName) return null
+  let imageUrl: string | null = null
+  if (rawImageUrl) {
+    try {
+      const resolvedUrl = new URL(rawImageUrl, baseUrl)
+      imageUrl = truncate(resolvedUrl.href, IMAGE_URL_MAX_CHARS)
```
[the remark this answers is not in the document]
Comment (BillLeoutsakosvl346) [reply apps/sim/app/api/link-preview/route.ts]: Changed image URL handling to drop URLs exceeding 2048 characters instead of truncating them. Updated test expectations accordingly.

File: apps/sim/app/api/link-preview/route.ts [hunk 3]
```diff
@@ -66,7 +82,7 @@ async function fetchPreview(url: string): Promise<LinkPreview> {
   if (!contentType.includes('text/html') && !contentType.includes('application/xhtml+xml')) {
     return null
   }
-  return parsePreview(await response.text())
+  return parsePreview(await response.text(), url)
```
[the remark this answers is not in the document]
Comment (BillLeoutsakosvl346) [reply apps/sim/app/api/link-preview/route.ts]: Extended `SecureFetchResponse` interface with a `url` field that tracks the final URL after redirects. Updated `parsePreview` to use `response.url` instead of the original request URL when resolving relative image URLs. Added test case to verify relative image URLs are resolved against the final redirected URL.

File: LINK_PREVIEW_IMAGE_CHANGES.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+- Verify authentication, rate limiting, and caching behavior
+
+Run tests with:
+```bash
+bun test apps/sim/app/api/link-preview/route.test.ts
+```
+
+## Future Enhancements
+
+The `ExternalLink` component (`apps/sim/app/workspace/[workspaceId]/home/components/message-content/components/chat-content/external-link.tsx`) currently displays title, description, and site name in a tooltip but does not display the image. The component could be enhanced to show the image thumbnail in the preview tooltip if desired.
```
[the remark this answers is not in the document]
Comment (BillLeoutsakosvl346) [reply LINK_PREVIEW_IMAGE_CHANGES.md]: Deleted LINK_PREVIEW_IMAGE_CHANGES.md from the repository.

File: apps/sim/app/api/link-preview/route.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (BillLeoutsakosvl346) [reply apps/sim/app/api/link-preview/route.ts]: Same fix as PRRT_kwDONmSNmM6T1IgM - oversized image URLs are now dropped rather than truncated with `...` suffix.

**Comment by BillLeoutsakosvl346** [comment]
@greptile

**Comment by BillLeoutsakosvl346** [comment]
@cursor review

File: apps/sim/app/api/link-preview/route.test.ts [hunk 5]
```diff
@@ -0,0 +1,875 @@
...
+            <meta property="og:image" content="https://example.com/image.jpg" />
+          </head>
+        </html>
+      `
+
+      inputValidationMockFns.mockSecureFetchWithValidation.mockResolvedValue(
+        Object.assign(new Response(html, {
+          status: 200,
+          headers: { 'content-type': 'text/html' },
+        }), { url: 'https://example.com' })
```
[the remark this answers is not in the document]
Comment (BillLeoutsakosvl346) [reply apps/sim/app/api/link-preview/route.test.ts]: Fixed the Response.url test mocks by creating a `createMockResponse` helper function that uses `Object.defineProperty` to properly set the read-only `url` property. All 19 occurrences of `Object.assign(new Response(...), { url: ... })` have been replaced with `createMockResponse(..., { ..., url: ... })`, which correctly allows the tests to exercise relative and redirect-based image URL resolution in `parsePreview`.

**Comment by BillLeoutsakosvl346** [comment]
@greptile

**Comment by BillLeoutsakosvl346** [comment]
@cursor review

File: apps/sim/lib/core/security/input-validation.server.ts [hunk 6]
```diff
@@ -462,6 +462,7 @@ export interface SecureFetchResponse {
   statusText: string
   headers: SecureFetchHeaders
   body: ReadableStream<Uint8Array> | null
+  url: string
```
[the remark this answers is not in the document]
Comment (BillLeoutsakosvl346) [reply apps/sim/lib/core/security/input-validation.server.ts:465]: Fixed by adding the required `url` property to all mock factories that return `SecureFetchResponse`-shaped objects:
- `mockSecureFetchResponse` in `apps/sim/app/api/tools/agiloft/attach/route.test.ts`
- `mockSecureFetchResponse` in `apps/sim/app/api/tools/agiloft/retrieve/route.test.ts`
- `mockSecureFetchResponse` in `apps/sim/app/api/tools/stt/route.test.ts`
- `fakeResponse` in `apps/sim/lib/knowledge/documents/utils.test.ts`

All mock factories now include `url: 'https://example.com'` to match the updated `SecureFetchResponse` interface.

File: apps/sim/app/api/link-preview/route.test.ts [hunk 7]
```diff
@@ -0,0 +1,911 @@
...
+      const data = await response.json()
+
+      expect(response.status).toBe(200)
+      expect(data.preview).toEqual({
+        title: 'Test Title',
+        description: null,
+        siteName: null,
+        imageUrl: null,
+      })
+    })
```
[the remark this answers is not in the document]
Comment (BillLeoutsakosvl346) [reply apps/sim/app/api/link-preview/route.test.ts:493]: Fixed by replacing `::invalid::` with `http://[::1` (incomplete IPv6 address) in the test case. The issue was that `new URL('::invalid::', 'https://example.com')` does not throw—it treats `::invalid::` as a valid relative path and resolves it to `https://example.com/::invalid::`. The new test value `http://[::1` is a malformed URL that actually causes `new URL()` to throw a TypeError, correctly exercising the error handling path.

**Comment by BillLeoutsakosvl346** [comment]
@greptile

**Comment by BillLeoutsakosvl346** [comment]
@cursor review
